src/Compilers/CSharp/Portable/Lowering/NullCheckRewriter.cs

```diff
@@ -0,0 +1,83 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using Microsoft.CodeAnalysis.CodeGen;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.CSharp;
+
+namespace Microsoft.CodeAnalysis.CSharp.Lowering
+{
+    class NullCheckRewriter : BoundTreeRewriterWithStackGuardWithoutRecursionOnTheLeftOfBinaryOperator
+    {
+        readonly MethodSymbol _method;
+        SyntheticBoundNodeFactory _fact;
+        public NullCheckRewriter(
+            MethodSymbol method,
+            SyntaxNode syntax,
+            TypeCompilationState compilationState,
+            DiagnosticBag diagnostics)
+        {
+            _method = method;
+            _fact = new SyntheticBoundNodeFactory(method, syntax, compilationState, diagnostics);
+        }
+
+        internal static BoundNode Rewrite(
+            BoundStatement body,
+            MethodSymbol method,
+            int methodOrdinal,
+            VariableSlotAllocator slotAllocatorOpt,
+            TypeCompilationState compilationState,
+            DiagnosticBag diagnostics) //,
+                                       //out IteratorStateMachine stateMachineType)
+        {
+            var rewriter = new NullCheckRewriter(method, body.Syntax, compilationState, diagnostics);
+            return rewriter.Visit(body);
+        }
+
+        public override BoundNode VisitBlock(BoundBlock node)
+        {
+            List<SourceParameterSymbolBase> toCheck = new List<SourceParameterSymbolBase>();
```

Comment:
>new List<SourceParameterSymbolBase>() [](start = 54, length = 37)

Why create a copy rather than looping over the original list in `AddNullChecksToBody`?